pkg/agent/proxy/route.go

```diff
@@ -0,0 +1,295 @@
+/*
+Copyright 2021 The Lynx Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/smartxworks/lynx/pkg/constants"
+)
+
+// NodeReconciler watch node and update route table
+type NodeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	StopChan    <-chan struct{}
+	updateMutex sync.Mutex
+}
+
+func GetNodeInternalIP(node corev1.Node) net.IP {
+	for _, item := range node.Status.Addresses {
+		if item.Type == corev1.NodeInternalIP {
+			return net.ParseIP(item.Address)
+		}
+	}
+	return nil
+}
+
+func GetRouteByDst(dst *net.IPNet) []netlink.Route {
+	var ret []netlink.Route
+	// List all route item in current node
+	routeList, err := netlink.RouteList(nil, unix.AF_INET)
+	if err != nil {
+		klog.Errorf("List route table error, err:%s", err)
+		return ret
+	}
+	for _, item := range routeList {
+		if item.Dst != nil && item.Dst.String() == dst.String() {
+			ret = append(ret, item)
+		}
+	}
+	return ret
+}
+func RouteEqual(r1, r2 netlink.Route) bool {
+	return ((r1.Dst == nil && r2.Dst == nil) ||
+		(r1.Dst != nil && r2.Dst != nil && r1.Dst.String() == r2.Dst.String())) &&
+		r1.Src.Equal(r2.Src) &&
+		r1.Gw.Equal(r2.Gw)
+}
+
+func UpdateRoute(nodeList corev1.NodeList, thisNode corev1.Node) {
```

Comment:
This function implementation is not quite explicit. Please add some comments for 'what' and 'why'.

Reply:
done